docs/build.py

```diff
@@ -6,249 +6,390 @@
+        # PDF download link in the PDF
+
+        sphinx_args = [
+            '-b',
+            'latex',
+            temp_directory,
+            latex_output_path,
+            '--fail-on-warning',
+            '--keep-going',
+            '-j',
+            f'{os.cpu_count()}'
+        ]
+
+        print('running sphinx to build latex files...')
+
+        sphinx_build(sphinx_args)
+
+        # Generate PDF
+        cmd(f'cd "{latex_output_path}" && latexmk -pdf "LVGL.tex"')
+
+        # Copy the result PDF to the main directory to make
+        # it available for the HTML build
+        shutil.copyfile(pdf_src_file, pdf_dst_file)
+
+        # cmd("cd out_latex && cp -f LVGL.pdf ../LVGL.pdf")
+
+        # add the PDF link so the HTML build will have it.
+        index_data = (
+            f'PDF version: :download:`LVGL.pdf <LVGL.pdf>`\n{index_data}'
+        )
+
+        with open(index_path, 'wb') as f:
+            f.write(index_data.encode('utf-8'))
+
+    def get_version():
+        path = os.path.join(project_path, 'lv_version.h')
+        with open(path, 'rb') as fle:
+            d = fle.read().decode('utf-8')
+
+        d = d.split('#define LVGL_VERSION_MAJOR', 1)[-1]
+        major, d = d.split('\n', 1)
+        d = d.split('#define LVGL_VERSION_MINOR', 1)[-1]
+        minor, d = d.split('\n', 1)
+
+        # d = d.split('#define LVGL_VERSION_PATCH', 1)[-1]
+        # patch, d = d.split('\n', 1)
+
+        return f'{major.strip()}.{minor.strip()}'
+
+    # I am not a big fan of runing subprocesses to run sphinx.
+    # Now we can run it directly without any issues.
+    sphinx_args = [
+        '-b',
+        'html',
+        html_src_path,
+        html_dst_path,
+        '-D',
+        f'version="{get_version()}"',
+        '--fail-on-warning',
+        '--keep-going',
+        '-E',
+        '-j',
+        f'{os.cpu_count()}'
+    ]
+
+    print('running sphinx to build html files...')
+    sphinx_build(sphinx_args)
+
+    if stdout.errors:
+        print('\n\n'.join(stdout.errors))
+    if stderr.errors:
+        print('\n\n'.join(stderr.errors))
+
+    if develop:
+        print('temp directory:', temp_directory)
+    else:
+        if not temp_path_exists:
             try:
-                os.rmdir(p)
+                shutil.rmtree(temp_directory)
             except OSError:
-                remove_folder = False
+                print(
+                    f'Unable to remove temporary directory "{temp_directory}"'
+                )
+        else:
+            print('temp directory:', temp_directory)
+
+    config_builder.cleanup()
 
-        return remove_folder
+    if stdout.errors or stderr.errors:
+        raise RuntimeError
 
-    iter_temp(temp_directory)
+    print('output path:', html_dst_path)
+    print('\nFINISHED!!')
 
-config_builder.cleanup()
 
-print('output path:', html_dst_path)
-print('\nFINISHED!!')
+# because sphinx can use multiple CPU cores when it compiles this requires
```

Comment:
It's how multiprocessing works when using python. The script gets run more than once because sphinx is being instructed to use multiple jobs which is what launches it more than once.

Reply:
I've added a line which generates a randomly named file upon the execution of `run()`, removed the `if __name__ == '__main__':` guard and called `build.py`. Only a single file was generated.

Could you explain me how to reproduce the "script being run more than once" scenario, please?

Reply:
Of course it is only going to make a single file upon execution of run. That's what it is supposed to do. Now if you make the file at the module level you will find that more than one file gets made. You can also just add a print statement at the module level as well so each time the thing executes there will be something printed out.

Reply:
Also this is only going to hold true if you are using a platform other than Windows.

Reply:
I'm using Linux, and still unable to recreate any scenario resembling what we are discussing.

I see no problem with the `if __name__ == '__main__':` guard as it's a common Python practice (although only recommended for modules, not standalone scripts). However I'd remove the comment as it might be confusing and the behavior it describes is not confirmed.

Reply:
read...

https://docs.python.org/3/library/multiprocessing.html

specifically this section

https://docs.python.org/3/library/multiprocessing.html#the-spawn-and-forkserver-start-methods

Now read this...

https://www.sphinx-doc.org/en/master/man/sphinx-build.html#cmdoption-sphinx-build-j

Reply:
delete the `if __name__ == '__main__:' line then run it and see what happens. It will run the entire document build multiple times in parallel, ultimately crashing as a result.

Reply:
I managed to get the following error on a virtual Windows machine:
![Screenshot from 2024-07-31 18-44-51](https://github.com/user-attachments/assets/9c66e6ea-a7e9-4fa1-bf6c-982032c761a0)
But the program exited successfully.

The situation is clearer now, my suggestion for the wording:
```
# we are invoking sphinx in multiprocessor mode (-j flag)
# which can cause waring messages / a crash under Windows
# (https://stackoverflow.com/a/50782028)
```

Reply:
This explains what is happening just fine.  There is no need to change it. It explains in detail what is taking place. 

```python
# because sphinx can use multiple CPU cores when it compiles this requires
# launching the script multiple times from different subprocesses. We only want
# the build script to run a single time and that is when the user runs the
# script. By checking the package name of this file to see if it is the "main"
# script puts in place a block so the build script only runs a single time
# even when the script gets run from multiple processes.
```

The description you are using is actually incorrect. `-j` is used to specify the number of processes (jobs) to launch, it has nothing to do with the number of processors. I can set that number to 30 if I wanted to and the computer could only have 6 cores (processors) and it would compile just fine. There would not be any real benefit from doing that performance wise. It would add a lot of overhead with no return on that investment. I have coded it so that the number of processes that is being run matches the logical processor count on the machine.